cpp/src/arrow/io/file.cc

```diff
@@ -199,6 +199,7 @@ class ReadableFile::ReadableFileImpl : public OSFile {
     RETURN_NOT_OK(Read(nbytes, &bytes_read, buffer->mutable_data()));
     if (bytes_read < nbytes) {
       RETURN_NOT_OK(buffer->Resize(bytes_read));
+      buffer->ZeroPadding();
```

Comment:
It seems error prone to have to remember to zero the padding. What would be the performance consequences (if any, realistically) of zeroing in Resize?

Reply:
Some things may be repeatedly resized if you don't know the exact size up front (e.g. in BinaryBuilder).

Reply:
Right, in the event that you shrank the buffer in `BinaryBuilder` when `Finish` is called, you would want to zero the padding there, too

Reply:
I guess if you shrink the memory then the memory you padded before goes unused

Reply:
The worst case is when you Reserve a huge amount of memory and then gradually Resize it, as the new elements come in. You would expect these operations to be virtually free while in reality you are zeroing everything between size and capacity on every call.

Reply:
Got it. OK, let's leave things as they are, then